Microsoft.data.sqlite

Felixmil opened the pull request:
Closes #3325 

Sister PRs:
- Core: https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/pull/2699
- MoBi: https://github.com/Open-Systems-Pharmacology/MoBi/pull/2163

# Description

This PR migrates PK-Sim from System.Data.SQLite to Microsoft.Data.Sqlite. This change enables macOS ARM64 support, improves modern .NET compatibility.

## Type of change

Please mark relevant options with an `x` in the brackets.

- [ ] Bug fix (non-breaking change which fixes an issue)

- [ ] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):

- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.

- [x] Other (please describe): Dependency migration.

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [ ] Unit tests
- [ ] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

Commit:
Migrate from System.Data.SQLite to Microsoft.Data.Sqlite

- Replace System.Data.SQLite.Core with Microsoft.Data.Sqlite 8.0.0
- Update DAS layer, NHibernate configuration, and database schema migrator
- Update all project files to use Microsoft.Data.Sqlite package
- Enables macOS ARM64 support and modern .NET compatibility

Commit:
Fix SnapshotRunner SQLite driver compatibility and error handling

- Update SessionFactoryProvider to use SQLite20Driver (compatible with Microsoft.Data.Sqlite)
- Add proper error handling and logging to SnapshotRunner for silent failures
- Move CloseSession call to outer finally block to ensure cleanup in all scenarios
- Add comprehensive test coverage for error scenarios
- Resolves issue where convertSnapshot function failed silently on macOS

Commit:
Update R dependency artifact creation to handle Microsoft.Data.Sqlite native libraries

Copy native SQLite libraries from runtimes folder to root before
creating the R dependencies artifact for Microsoft.Data.Sqlite compatibility.

Commit:
Merge upstream/V13 into Microsoft.Data.Sqlite branch

- Resolved merge conflicts in:
  - SnapshotRunner.cs: Removed duplicate logging line
  - PKSim.Infrastructure.csproj: Kept Microsoft.Data.Sqlite, updated to V13 package versions
  - PKSim.csproj: Kept Microsoft.Data.Sqlite, updated to V13 versions, kept PostBuild target
  - PKSim.UI.Tests.csproj: Kept Microsoft.Data.Sqlite, updated to V13 versions
- Deleted files removed in upstream/V13:
  - PKSimRDependencyResolution.csproj
  - PKSim.Matlab.Tests.csproj

Commit:
Remove leftover SQLite interop properties from System.Data.SQLite migration

These properties (ContentSQLiteInteropFiles, CopySQLiteInteropFiles, etc.)
were used by System.Data.SQLite.Core to handle native library deployment.
Microsoft.Data.Sqlite handles native libraries through NuGet package assets
automatically, so these properties are no longer needed.

Commit:
Merge latest upstream/V13 into Microsoft.Data.Sqlite branch

- Resolved conflicts keeping Microsoft.Data.Sqlite migration
- Removed Foreign Keys connection string parameter (Microsoft.Data.Sqlite has foreign keys enabled by default)
- Kept manual data reading implementation in FillDataTable for Microsoft.Data.Sqlite compatibility
- Preserved additional error handling test classes

Commit:
Disable foreign keys in Microsoft.Data.Sqlite connection string

Felixmil commented on .github/workflows/build-nightly_v13.yml:
```diff
@@ -96,11 +106,25 @@ jobs:
           pushd .
           cd src\PKSimRDependencyResolution\bin\Debug\${{env.TARGET_FRAMEWORK}}
-          7z.exe a pk-sim-r-dependencies.zip -x@"excludedFiles.txt" * -xr!runtimes
+          @("win-x64", "linux-x64", "osx-arm64") | ForEach-Object {
```
osx-x64 notsupported.

Felixmil commented on README.md:
```diff
@@ -1,32 +1,12 @@
 
 ## Code Status
-[![Build status](https://img.shields.io/github/actions/workflow/status/Open-Systems-Pharmacology/PK-Sim/build-and-test.yml?logo=nuget&label=Build%20status)](https://github.com/Open-Systems-Pharmacology/PK-Sim/actions/workflows/build-and-test.yml)
+[![Build status](https://img.shields.io/github/actions/workflow/status/Open-Systems-Pharmacology/PK-Sim/nightly-badge.yml?logo=GitHub&label=Build%20status)](https://github.com/Open-Systems-Pharmacology/PK-Sim/actions/workflows/build-nightly_12.2.yml)
```
Not in the scope of this PR

Commit:
Fix duplicate error logging in SnapshotRunner

Commit:
Use Count property instead of Count() LINQ extension

Felixmil commented on src/PKSim.Core/Services/InteractiveSimulationRunner.cs:
```diff
@@ -51,7 +61,7 @@ public InteractiveSimulationRunner(
          {
             CheckForNegativeValues = true,
-            RaiseEvents = true,
+            RaiseEvents = true, //This is always true, but still worth checking before rising an event since it can change.
```
Not in the scope of this PR

Commit:
Merge upstream/V13 into Microsoft.Data.Sqlite branch

Commit:
Revert "Use Count property instead of Count() LINQ extension"

This reverts commit 14b5b4367c66884a09a8a29a15b1bf5a37921f52.

rwmcintosh commented on .github/workflows/build-nightly_v13.yml:
```diff
@@ -105,7 +105,16 @@ jobs:
+          }
+          if (Test-Path "runtimes") {
+            Remove-Item -Recurse -Force "runtimes"
+          }
```
We should do this file copy the same way we do that with other native dll that have to be manually placed in the output directory.

rwmcintosh replied:
using `GeneratePathProperty="true"` and a `Content Include`

Felixmil replied:
I tried but now we get this during build:
```
C:\Program Files\dotnet\sdk\9.0.306\Microsoft.Common.CurrentVersion.targets(5394,5): error MSB3030: Could not copy the file "C:\Users\runneradmin\.nuget\packages\microsoft.data.sqlite\8.0.0\runtimes\linux-x64\native\libe_sqlite3.so" because it was not found. [D:\a\PK-Sim\PK-Sim\src\PKSim.R\PKSim.R.csproj]
C:\Program Files\dotnet\sdk\9.0.306\Microsoft.Common.CurrentVersion.targets(5394,5): error MSB3030: Could not copy the file "C:\Users\runneradmin\.nuget\packages\microsoft.data.sqlite\8.0.0\runtimes\win-x64\native\e_sqlite3.dll" because it was not found. [D:\a\PK-Sim\PK-Sim\src\PKSim.R\PKSim.R.csproj]
C:\Program Files\dotnet\sdk\9.0.306\Microsoft.Common.CurrentVersion.targets(5394,5): error MSB3030: Could not copy the file "C:\Users\runneradmin\.nuget\packages\microsoft.data.sqlite\8.0.0\runtimes\osx-arm64\native\libe_sqlite3.dylib" because it was not found. [D:\a\PK-Sim\PK-Sim\src\PKSim.R\PKSim.R.csproj]
```

rwmcintosh replied:
I removed this section completely. I thought I had already done it but I guess not. In V13 we won't use this dependency artifact.

rwmcintosh commented on src/PKSim.CLI.Core/Services/SnapshotRunner.cs:
```diff
@@ -89,22 +89,29 @@ private async Task startSnapshotRun(IReadOnlyList<FileMap> fileMaps, Func<FileMa
+         catch (Exception ex)
+         {
+            _logger.AddError($"Error processing snapshot file '{file.SnapshotFile}': {ex.Message}");
+            throw;
```
These changes aren't related to SQLite change.  The `AddError` on 108 will duplicate the log entry because in the caller there's already a catch and log.

Felixmil replied:
When converting snapshot to other format from R, I had an uncaught error. So I added tests cases. However, you are right, it is out of scope for now. Will revert changes related to SnapshotRunner

Commit:
Revert SnapshotRunner changes - out of scope for SQLite migration

Felixmil commented on tests/PKSim.Tests/CLI/SnapshotRunnerSpecs.cs:
```diff
@@ -172,4 +173,91 @@ public void should_generate_the_project_from_snapshot()
+         _runOptions.ExportMode = SnapshotExportMode.Project;
+         _runOptions.InputFolder = _inputFolder;
+         _runOptions.OutputFolder = _outputFolder;
+         _snapshotFile = Path.Combine(_inputFolder, $"{_fileName}{Constants.Filter.JSON_EXTENSION}");
```
What does this mean ?

Yuri05 replied:
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/Open-Systems-Pharmacology/PK-Sim/pull/3323#discussion_r2490217059)

Yuri05 replied:
no idea tbh. Does not make much sense for me.
let's see what Copilot will propose in its PR

Yuri05 replied:
> no idea tbh. Does not make much sense for me. let's see what Copilot will propose in its PR

Ok, this did not work out.
I'll set these suggestions to resolved, not all Copilot proposals are reasonable.

Commit:
Aligns with how other native DLLs are handled

- Add Content Include items to copy SQLite native libraries to output directory
- Simplify workflow by removing PowerShell copy script, use -xr!runtimes flag

msevestre commented on src/PKSim.Infrastructure/ORM/DAS/DAS.cs:
```diff
@@ -153,8 +159,9 @@ public bool Connect(string dbName, string workgroupDb, string user, string passw
-                  sProvider = "System.Data.SQLite";
+                  sProvider = "Microsoft.Data.Sqlite";
                   connBuilder.Add("Data Source", dbName);
+                  connBuilder.Add("Foreign Keys", "False");
```
why false? Shoudn't that be TRUE?

rwmcintosh replied:
This is a behavior difference between the old data access and new. The test will fail if we enable foreign keys

When_deleting_the_reference_to_another_template

[I'll open an issue ](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/3331)

msevestre replied:
I don’t  Think we should change the behaviour of the app just because a test fail.

rwmcintosh replied:
Well I was thinking it's a difference in the default connection parameters but maybe not.

We should find if the test behaviour is different

rwmcintosh replied:
The default connection parameters for `System.Data.SQLite` does not enable foreign key enforcement, so for this PR we can disable it explicitly. @Yuri05 will enter an issue for the template database delete.

Yuri05 replied:
> The default connection parameters for `System.Data.SQLite` does not enable foreign key enforcement, so for this PR we can disable it explicitly. @Yuri05 will enter an issue for the template database delete.

https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/3332

rwmcintosh committed:
Update to use NHibernate.Extensions.SQLite and new core

rwmcintosh commented:
This won't build until we get a core merge, but this now works and only uses `Microsoft.Data.SQLite` where up to now it was using a hybrid of both.

msevestre commented on src/PKSim.Infrastructure/ORM/DAS/DAS.cs:
```diff
@@ -19,6 +19,12 @@ public class DAS : IDisposable
+      public DAS()
+      {
+         // Initialize the provider factory to Microsoft.Data.Sqlite by default
+         _providerFactory = Microsoft.Data.Sqlite.SqliteFactory.Instance;
```
why do we use the full namespace here? 
Also why do we instantiate in constgructor ?
This DAS the way it's written does not work with anything else but sqlite anymore. Should we change the code to simplify?

rwmcintosh replied:
removed. It's initialized elsewhere.

rwmcintosh commented on src/PKSim.Infrastructure/ORM/DAS/DAS.cs:
```diff
@@ -508,10 +515,36 @@ public void FillDataTable(DASDataTable dataTable, string sql)
          try
          {
-            using (var adapter = _providerFactory.CreateDataAdapter())
+            // Manual data reading to avoid constraint issues with DataTable.Load()
```
This change is a bit of a mystery to me. Are there some constraints that were ignored before

Yuri05 replied:
The **views** have no constraints at all. Only the **tables** have, but they are never accessed by PK-Sim **directly** anyways - only via the views.

rwmcintosh replied:
I refactored the System.Data.SQLite code to use `dataTable.Load` and there are no constraints for this view

"SELECT * FROM VIEW_POPULATIONS"

<img width="639" height="348" alt="Image" src="https://github.com/user-attachments/assets/7fe1205a-431f-470d-bd16-373bddb7c97e" />

With the new Micorosoft.Data.SQLite a different story

<img width="495" height="424" alt="Image" src="https://github.com/user-attachments/assets/49467af9-9199-4960-bdf3-23cec69df0ed" />

Several rows show a constraint violation.

<img width="599" height="217" alt="Image" src="https://github.com/user-attachments/assets/c6f24f52-52bd-4f28-828d-23cd014d8cf4" />

The Integration tests show this failure with other views too. I don't know all of them.

msevestre replied:
What is this code doing in the first place?

rwmcintosh replied:
dataTable.Load method
<img width="564" height="128" alt="image" src="https://github.com/user-attachments/assets/699391ea-b117-47d1-befa-10e40576641f" />

rwmcintosh replied:
It's because the constraints are built based on the underlying tables. In this example `tab_species` does have a unique key on the species column since it's the primary key for that table.

The manual update code does not analyze the schema and so does not have this problem. It's also possible that the System.Data.SQLite also could not determine the underlying tables.

msevestre commented on src/PKSim.Infrastructure/ORM/DAS/DAS.cs:
```diff
@@ -508,10 +515,36 @@ public void FillDataTable(DASDataTable dataTable, string sql)
          {
-            using (var adapter = _providerFactory.CreateDataAdapter())
+            // Manual data reading to avoid constraint issues with DataTable.Load()
+            using (var reader = cmd.ExecuteReader())
```
what are we doing here? there is no adapater for this

rwmcintosh replied:
<img width="521" height="151" alt="image" src="https://github.com/user-attachments/assets/d6cbb43d-c9f2-4013-baa0-a804c83d57dc" />

Yuri05 replied:
Would **this** work? Replace `dataTable.Load(reader)` with
```
dataTable.BeginLoadData();
dataTable.Load(reader);

// Remove any inferred constraints so EndLoadData won’t validate them
dataTable.Constraints.Clear();
dataTable.PrimaryKey = Array.Empty<DataColumn>();
foreach (DataColumn col in dataTable.Columns)
{
    col.Unique = false;
}

dataTable.EndLoadData();
```

As far as I understood, [`dataTable.BeginLoadData()`](https://learn.microsoft.com/en-us/dotnet/api/system.data.datatable.beginloaddata?view=net-8.0) turns all constraints/index checks off. 
Then clear all of them before `dataTable.EndLoadData();` turns constraints/index checks on again.

rwmcintosh replied:
You can ignore them in a few ways. Even then there's a few failing tests that I'll look into today.

rwmcintosh replied:
It's not only constraints that cause a different behavior. The DataTable.Load method will analyze the underlying tables and apply the keys to the data table. In one case for `FlatRateObjectPathRepository` the compound key is on 4 columns. Where we usually get 2 object paths, we only get 1 because the second is ignored as the data table is filled.

rwmcintosh committed:
create a custom data adapter

rwmcintosh commented on src/PKSim.Infrastructure/ORM/DAS/DAS.cs:
```diff
@@ -508,18 +515,18 @@ public void FillDataTable(DASDataTable dataTable, string sql)
-               adapter.SelectCommand = cmd;
-               adapter.Fill(dataTable);
-            }
+            using var adapter = new DASSQLAdapter();
+            adapter.SelectCommand = cmd;
+            adapter.Fill(dataTable);
```
Could it be this simple?

rwmcintosh commented on src/PKSim.Infrastructure/Services/SessionFactoryProvider.cs:
We are having a problem here with an exception thrown and ignored within NHibernate.

msevestre replied:
We might to upgrade n hibernate 

Another option is to move away from Nhiberbste completely and use what we use for the working journal

rwmcintosh replied:
Upgrading won't help. It's not compatible with Microsoft.Data.SQLite

msevestre replied:
Not at all. ? Why not. Isn’t the DLL just an abstraction to access the DB

msevestre replied:
To be honest , nhiberbate is overkill for what we are doing. But migrating away from it won’t be a walk in the park. Using dapper will probably be easy enough considering that we barely ever write our own sql

rwmcintosh replied:
We are going to remove the automatic schema update through nhibernate. I did not write this down, but I have added the schema to source control

rwmcintosh committed:
add project schema dump as source

rwmcintosh committed:
do not automatically update schema from nhibernate

rwmcintosh commented on src/PKSim.Infrastructure/Services/SessionFactoryProvider.cs:
```diff
@@ -22,8 +22,6 @@ public ISessionFactory InitalizeSessionFactoryFor(string dataSource)
       public ISessionFactory OpenSessionFactoryFor(string dataSource)
       {
          var cfg = createSqlLiteConfigurationFor(dataSource);
-         var update = new SchemaUpdate(cfg);
-         update.Execute(useStdOut: false, doUpdate: true);
```
This won't work with Microsoft.Data.SQLite. Based on their public GitHub discussions, it won't work anytime soon.

We will have to manage migrations manually in the project database. That will also be required in MoBi.

rwmcintosh replied:
We can write it in the [developer docs](https://github.com/Open-Systems-Pharmacology/developer-docs/issues/71).

rwmcintosh committed:
remove constructor init of factory

rwmcintosh committed:
foreign keys enabled

rwmcintosh committed:
enable foreign keys

rwmcintosh committed:
update core

rwmcintosh commented on src/PKSim.Infrastructure/Serialization/DatabaseSchemaMigrator.cs:
```diff
@@ -19,7 +19,7 @@ public class DatabaseSchemaMigrator : IDatabaseSchemaMigrator
       {
          var path = fileFullPath.ToUNCPath();
-         using (var sqlLite = new SQLiteConnection($"Data Source={path}"))
+         using (var sqlLite = new SqliteConnection($"Data Source={path};Foreign Keys=False"))
```
removed

rwmcintosh committed:
pr feedback

rwmcintosh committed:
undo

rwmcintosh commented on tests/PKSim.Tests/Infrastructure/SessionFactoryProviderSpecs.cs:
```diff
@@ -1,44 +1,50 @@
+      [Observation]
+      public void should_be_able_to_open_a_session()
+      {
+         using (var session = _sessionFactory.OpenSession())
```
removed

rwmcintosh commented on tests/PKSim.Tests/IntegrationTests/LazyLoadingIntegrationTests.cs:
```diff
@@ -1,66 +1,62 @@
+         {
+            var fromDb = session.Load<IndividualMetaData>(_individualMetaData.Id);
+            NHibernateUtil.IsInitialized(fromDb.Content).ShouldBeFalse();
+            var content = fromDb.Content.Data;
```
we need this for side effect

rwmcintosh committed:
the default connection did not enforce foreign keys

rwmcintosh committed:
update core

rwmcintosh commented:
I checked the file size differences against the V13 branch by loading a large snapshot file and saving that to pksim5 file.

<img width="828" height="527" alt="image" src="https://github.com/user-attachments/assets/3a9135c6-4ff9-4861-9d2b-c2a3f2442570" />

They are very close.

rwmcintosh commented:
After removing all the simulations as well.

<img width="832" height="532" alt="image" src="https://github.com/user-attachments/assets/4ccb4373-54fb-4728-b349-3db66169c318" />

msevestre commented on src/PKSim/PKSim.csproj:
```diff
@@ -91,12 +91,12 @@
+    <Reference Include="System.Windows.Forms" />
+  </ItemGroup>
+
+  <Target Name="PostBuild" AfterTargets="PostBuildEvent" Condition="'$(NCrunch)' != '1'">
```
what is this stuff?

rwmcintosh replied:
Not sure where that comes from. It should not be needed.

msevestre reviewed (approved):
looks good to me with one question regarding weird command in one project file. Approving but please review just in case this is not intended

rwmcintosh committed:
remove postbuild and unused references
